feat: add support for per donation form Stripe accounts #4591

mehul0810 opened the pull request:
## Description
This PR resolves #4591 

## Affects
Majority of the code under `includes > gateways > stripe` and `assets` related to Stripe.

## What to test
Check the acceptance criteria on the issue and try to break the functionality with all the payment methods supported by Stripe.

## Screenshots:
**This is how all the Stripe accounts will look like when connected**
![image](https://user-images.githubusercontent.com/1852711/79984437-ce6f1980-84c6-11ea-91cb-115b4322147e.png)

**Register Apply Pay domain for each account**
![image](https://user-images.githubusercontent.com/1852711/79984641-1b52f000-84c7-11ea-8afd-e54a56ae9b50.png)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

JasonTheAdams commented on assets/src/js/admin/stripe-admin.js:
```diff
@@ -27,13 +27,145 @@ window.addEventListener( 'DOMContentLoaded', function() {
 	giveStripeJsonFormattedTextarea( stripeStylesComplete );
 	giveStripeJsonFormattedTextarea( stripeCustomFonts );
 
+	if ( null !== perAccountEdits ) {
```
When doing direct DOM manipulation it is especially important to leave comments to document what's going on, as it is by no means obvious what's happening. Can you please go through this and leave a comment here and there to indicate what each hunk of code is doing?

mehul0810 replied:
Added detailed comments

JasonTheAdams commented on assets/src/js/frontend/give-stripe-becs.js:
```diff
@@ -30,6 +24,22 @@ document.addEventListener( 'DOMContentLoaded', function( e ) {
 	Array.prototype.forEach.call( formWraps, function( formWrap ) {
 		const formElement = formWrap.querySelector( '.give-form' );
 
+		// Bailout, if `form_element` is null.
```
Why would we bail? Is this an expected scenario? Or should we throw an error? If this expected please update the comment to explain what that scenario is and why it's ok to return at this point.

mehul0810 replied:
Added detailed comments as it is expected scenario

JasonTheAdams commented on assets/src/js/frontend/give-stripe-sepa.js:
```diff
@@ -30,6 +24,22 @@ document.addEventListener( 'DOMContentLoaded', function( e ) {
 	Array.prototype.forEach.call( formWraps, function( formWrap ) {
 		const formElement = formWrap.querySelector( '.give-form' );
 
+		// Bailout, if `form_element` is null.
```
Same here, please explain this.

mehul0810 replied:
Added detailed comments as it is expected scenario

JasonTheAdams commented on includes/admin/give-metabox-functions.php:
```diff
@@ -370,6 +374,43 @@ class="give-select-chosen give-chosen-settings"
+ * Output a label field
+ * Note: only for internal use.
+ *
+ * @param array $field         {
+ *                              Optional. Array of text input field arguments.
+ *
+ * @type string $id            Field ID. Default ''.
+ * @type string $wrapper_class CSS class to use for wrapper of input field. Default ''.
+ * @type string $value         Value of input field. Default ''.
+ * @type string $title         Text to display. HTML supported Default ''.
+ * @type string $before_field  Text/HTML to add before input field. Default ''.
+ * @type string $after_field   Text/HTML to add after input field. Default ''.
+ * }
```
Do we have a Field class we can use, instead?

mehul0810 replied:
No, we don't have any Field class in Give core. I have added an additional field support in Give core using the way the support for fields is already implemented.

JasonTheAdams replied:
I see. We'll want to refactor this at some point, then. If an object has a predictable shape, then it should be a class rather than a mutational array — which has no constraints on its shape. Don't worry about it, now, but keep this in mind down the road, please. 😃

JasonTheAdams commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -686,95 +697,216 @@ public function stripe_modal_checkout_status( $status = 'enabled' ) {
 
 		/**
-		 * Connect button to connect with Stripe account.
+		 * Stripe Account Manager Field.
```
Please provide a description for this. Does this render the Stripe Account Manager Field? What does this do?

mehul0810 replied:
Added description

JasonTheAdams commented on includes/gateways/stripe/includes/payment-methods/class-give-stripe-checkout.php:
```diff
@@ -306,11 +306,12 @@ public function process_checkout( $donation_id, $data ) {
 
 			// Redirect to show loading area to trigger redirectToCheckout client side.
-			wp_safe_redirect(
+			wp_redirect(
```
Any reason we wouldn't continue using `wp_safe_redirect` here?

mehul0810 replied:
Reverted the change as not an intended change

mehul0810 commented:
@DevinWalker @ravinderk @JasonTheAdams 

**Please find the testing videos as below:**
1. https://www.loom.com/share/a5b12a286a6e4ae781836e23aadb81cb
2. https://www.loom.com/share/885e1b545af7416d8754ece18e7c8aca

DevinWalker commented on assets/src/js/frontend/give-donations.js:
```diff
@@ -250,7 +250,7 @@ jQuery( function( $ ) {
 			}
 		}
-
+console.log(value_now);
```
Remove

DevinWalker commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +430,102 @@ function give_stripe_show_currency_notice() {
+		];
+		wp_send_json_success( $success_args );
+	} else {
+		wp_send_json_error( [ 'message' => __( 'Unable to update account name. Please contact administrator.', 'give' ) ] );
```
Change to "Unable to update account name. Please contact support."

DevinWalker commented on includes/admin/upgrades/upgrade-functions.php:
```diff
@@ -142,6 +142,10 @@ function give_do_automatic_upgrades() {
 			give_v2511_upgrades();
 			$did_upgrade = true;
+
+		case version_compare( $give_version, '2.6.3', '<' ):
```
@mehul0810 this is a pretty massive update to Stripe so we're going to hold off releasing this until `2.7.0` to give adequate time for testing and documentation - please update the version number throughout to reflect that as well as the upgrade process.

mehul0810 replied:
@DevinWalker Can you also let me know what version you are planning to release along with Give 2.7.0 for Recurring, Stripe Premium, and Fee Recovery?

DevinWalker commented on includes/gateways/stripe/includes/payment-methods/class-give-stripe-checkout.php:
```diff
@@ -363,9 +366,15 @@ public function redirect_to_checkout() {
-					const stripe = Stripe( '<?php echo $publishable_key; ?>', {
+					let stripe = {};
+
+					stripe = Stripe( '<?php echo $publishable_key; ?>' );
```
Have you tested the per-form account functionality with multiple donation forms on a page using different Stripe accounts? As well, testing the form grid functionality to verify it works as expected? Those are going to be very popular use cases.

mehul0810 replied:
@DevinWalker I have noticed a couple of minor issues which I fixed with the latest commits. Thanks for pointing out these scenarios.

DevinWalker commented:
@mehul0810 please add a "Cancel" button that appears when I click "Edit" so that if I clicked accidentally I can return to the previous state:

![2020-04-29_14-40-54](https://user-images.githubusercontent.com/1571635/80649915-bdbf2600-8a27-11ea-8769-38cddcec038b.png)

mehul0810 commented:
@DevinWalker Please find the updates below:

**Regarding "Add Cancel Button"**
- Added

**Regarding "Change "Manual" and "Connect" wording"**
- Updated

**Regarding "Improve Donation View"**
- I did some research on this point and it seems that if a user connects a Stripe account "Account A" and then processes a donation with "Account A" and then attach another Stripe account "Account B" to the same form then it will point to wrong information in that case.

Another thing, I thought is that there is no way to detect that manually connected account belongs to which specific account. For example, we can get an account id (i.e. acct_xxxxxx) for a connected account. But, we can't get it for a manual Stripe account. So, what do you propose with this scenario?

Regarding "Copy Style from Manage Accounts to Google / Apple Pay Table"
- I have updated the styling for Google/Apple Pay table. Please see screenshots below:

![image](https://user-images.githubusercontent.com/1852711/80975977-7c23e600-8e40-11ea-9856-8e3a46cacadf.png)

![image](https://user-images.githubusercontent.com/1852711/80975996-847c2100-8e40-11ea-9b0d-27b465558b5d.png)

Let me know if you have any questions.

DevinWalker commented:
@mehul0810 thanks for the update. Here are my thoughts on your question regarding the donation details view.

> I did some research on this point and it seems that if a user connects a Stripe account "Account A" and then processes a donation with "Account A" and then attach another Stripe account "Account B" to the same form then it will point to wrong information in that case.

In the scenario you describe above the donation data would still be relevant because at the time of the donation the donation form was set up to process donations with "Account A". Even if an admin goes in at a later date and changes the donation form to "Account B" the past donation was still accepted by "Account A". The data should be saved to the donation meta, not pulled from the currently connected account, because it the information should be preserved as a record.

> Another thing, I thought is that there is no way to detect that manually connected account belongs to which specific account. For example, we can get an account id (i.e. acct_xxxxxx) for a connected account. But, we can't get it for a manual Stripe account. So, what do you propose with this scenario?

My thought here is that if an account is using API keys it still has an account name associated with it. So the donation record would read:

"Donation accepted with Stripe account "Account A" using Manual API keys."

If connected the record would read:

"Donation accepted with Stripe account "Account B" _acct_1294183_ using Stripe Connect."

Let me know if you have any questions about the above.

bradp commented on includes/admin/upgrades/upgrade-functions.php:
```diff
@@ -3593,3 +3597,49 @@ function give_v2511_upgrades() {
+	// Process, only when there is no Stripe accounts stored.
+	if (
+		is_array( $stripe_accounts ) &&
+		count( $stripe_accounts ) === 0 &&
```
Should be yoda conditional.

bradp commented on includes/admin/upgrades/upgrade-functions.php:
```diff
@@ -3593,3 +3597,49 @@ function give_v2511_upgrades() {
+ * @return void
+ */
+function give_v270_upgrades() {
+
```
Remove extra line break.

bradp commented on includes/gateways/stripe/includes/actions.php:
```diff
@@ -33,15 +33,19 @@
 
-	<fieldset id="give_cc_fields" class="give-do-validate">
+	<fieldset
+		id="give_cc_fields_<?php echo $id_prefix; ?>"
```
`$id_prefix`  needs to be escaped.

bradp commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -41,17 +41,47 @@ function give_stripe_connect_save_options() {
+
+	if ( is_array( $stripe_accounts ) && count( $stripe_accounts ) === 0 ) {
+		// Set the connect details of Stripe account as default Stripe account.
+		give_update_option( 'give_stripe_connected', $get_vars['connected'] );
```
These values should be sanitized.

bradp commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +430,102 @@ function give_stripe_show_currency_notice() {
+		$account_name = ! empty( $get_data['account_name'] ) ? $get_data['account_name'] : false;
+
+		// Disconnect Stripe Account.
+		give_stripe_disconnect_account( $account_name );
```
This value should be sanitized.

ravinderk replied:
@brad already sanitized at beginning of this function.

mehul0810 replied:
@bradp The complete `$get_data` array is sanitized already so no need to sanitize it separately.

bradp commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +430,102 @@ function give_stripe_show_currency_notice() {
+ * @return void
+ */
+function give_stripe_update_account_name() {
+
```
Remove extra line break.

bradp commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +430,102 @@ function give_stripe_show_currency_notice() {
+ * @return void
+ */
+function give_stripe_set_account_default() {
+
```
Remove extra line break.

bradp commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +430,102 @@ function give_stripe_show_currency_notice() {
+ * @return void
+ */
+function give_stripe_disconnect_connect_stripe_account() {
+
```
Remove extra line break.

ravinderk replied:
@mehul can you fix this.

bradp commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +430,102 @@ function give_stripe_show_currency_notice() {
+	$is_updated = give_update_option( '_give_stripe_default_account', $account_slug );
+
+	if ( $is_updated ) {
+		wp_send_json_success();
```
Is there any data we want to pass in the success response?

mehul0810 replied:
@bradp No, we don't need any data to be passed in success as of now.

bradp commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+				'type'          => 'radio',
+				'default'       => $first_account_key,
+				'options'       => give_stripe_get_account_options(),
+				'wrapper_class' => $form_account ? 'give-stripe-per-form-default-account' : 'give-stripe-per-form-default-account give-hidden',
```
`$form_account` should be escaped.

ravinderk replied:
@mehul0810 Can we fix this?

bradp commented on includes/gateways/stripe/includes/admin/admin-helpers.php:
```diff
@@ -151,7 +152,7 @@ function give_stripe_connect_button() {
 			'return_url'            => rawurlencode( admin_url( 'edit.php?post_type=give_forms&page=give-settings&tab=gateways&section=stripe-settings' ) ),
 			'website_url'           => get_bloginfo( 'url' ),
-			'give_stripe_connected' => ! empty( $connected ) ? '1' : '0',
+			'give_stripe_connected' => '0',//! empty( $connected ) ? '1' : '0',
```
Remove old code in comment.

bradp commented on includes/gateways/stripe/includes/admin/admin-helpers.php:
```diff
@@ -166,24 +167,32 @@ function give_stripe_connect_button() {
  */
-function give_stripe_disconnect_url() {
+function give_stripe_disconnect_url( $account_id = '', $account_name = '' ) {
+
```
Remove extra line break.

ravinderk replied:
@mehul can we fix this?

bradp commented on includes/gateways/stripe/includes/admin/admin-helpers.php:
```diff
@@ -204,3 +213,22 @@ function give_stripe_connect_delete_options() {
+ * @return string
+ */
+function give_stripe_connection_type_names( $slug = 'connect' ) {
+
```
Remove extra line break.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
 		 */
-		public function stripe_connect_field( $value, $option_value ) {
+		public function stripe_account_manager_field( $field, $option_value ) {
+
```
Remove extra line break.

ravinderk replied:
@mehul can we fix this?

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+			$default_account = '';
+
+			// Set account as default.
+			if ( is_array( $stripe_accounts ) && count( $stripe_accounts ) === 1 ) {
```
Should be yoda conditional.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+			}
+
+			$site_url            = get_site_url();
+			$modal_title         = __( '<strong>You are connected! Now this is important: Please configure your Stripe webhook to finalize the setup.</strong>', 'give' );
```
HTML should not be passed to translation function.

bradp replied:
Because translated strings are user-defined, the translated string should be escaped.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+			$modal_title         = __( '<strong>You are connected! Now this is important: Please configure your Stripe webhook to finalize the setup.</strong>', 'give' );
+			$modal_first_detail  = sprintf(
+				'%1$s %2$s',
+				__( 'In order for Stripe to function properly, you must add a new Stripe webhook endpoint. To do this please visit the <a href=\'https://dashboard.stripe.com/webhooks\' target=\'_blank\'>Webhooks Section of your Stripe Dashboard</a> and click the <strong>Add endpoint</strong> button and paste the following URL:', 'give' ),
```
Because translated strings are user-defined, the translated string should be escaped.

ravinderk replied:
@mehul0810 Can we fix this?

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+				__( 'In order for Stripe to function properly, you must add a new Stripe webhook endpoint. To do this please visit the <a href=\'https://dashboard.stripe.com/webhooks\' target=\'_blank\'>Webhooks Section of your Stripe Dashboard</a> and click the <strong>Add endpoint</strong> button and paste the following URL:', 'give' ),
+				"<strong>{$site_url}?give-listener=stripe</strong>"
+			);
+			$modal_second_detail = __( 'Stripe webhooks are required so GiveWP can communicate properly with the payment gateway to confirm payment completion, renewals, and more.', 'give' );
```
Because translated strings are user-defined, the translated string should be escaped.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
-							class="stripe-btn-disabled"
+					<div id="give-stripe-account-manager-errors"></div>
+					<div id="give-stripe-account-manager-description">
+						<h2><?php _e('Manage Your Stripe Accounts'); ?></h2>
```
Missing spaces before/after the `'`.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+					<div id="give-stripe-account-manager-errors"></div>
+					<div id="give-stripe-account-manager-description">
+						<h2><?php _e('Manage Your Stripe Accounts'); ?></h2>
+						<p class="give-field-description"><?php _e('In this section you can connect one or multiple Stripe accounts. All donation forms will use the "Default Account" attached unless you specify otherwise in the specific donation form\'s settings. Connecting multiple accounts allows you to create donation campaigns for specific accounts rather than just one.', 'give'); ?></p>
```
Missing spaces before/after the `'`.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+									$disconnect_message = ( 'connect' === $details['type'] ) ?
+										sprintf(
+											__( 'Are you sure you want to disconnect GiveWP from Stripe? If disconnected, this website and any others sharing the same Stripe account (%1$s) that are connected to GiveWP will need to reconnect in order to process payments.', 'give' ),
+											$stripe_account_id
```
Because translated strings are user-defined, the translated string should be escaped.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+											__( 'Are you sure you want to disconnect GiveWP from Stripe? If disconnected, this website and any others sharing the same Stripe account (%1$s) that are connected to GiveWP will need to reconnect in order to process payments.', 'give' ),
+											$stripe_account_id
+										) :
+										__( 'Are you sure you want to disconnect GiveWP from Stripe?', 'give' );
```
Because translated strings are user-defined, the translated string should be escaped.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+												<?php echo give_stripe_convert_slug_to_title( $name ); ?>
+											</span>
+											<span class="give-stripe-account-edit">
+												<a class="give-stripe-account-edit-name" href="#"><?php echo __( 'Edit', 'give' ); ?></a>
```
Use `e__()`

bradp replied:
oops, I meant `_e()`

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+													href="#"
+													data-account="<?php echo $name; ?>"
+												><?php _e( 'Update', 'give' ); ?></a>
+												<a class="give-stripe-account-cancel-name give-hidden" href="#"><?php echo __( 'Cancel', 'give' ); ?></a>
```
Use `e__()`

bradp replied:
oops, I meant `_e()`

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+											</span>
+										</div>
+										<div class="give-stripe-account-actions">
+											<span class="give-stripe-account-type-description give-field-description"><?php _e( 'Connection Method:', 'give' ); ?></span>
```
These values echoed into the HTML need to have `esc_attr()` applied to them.

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+							<?php } ?>
+						</div>
+						<div class="give-stripe-account-manager-add-section">
+							<h3><?php echo __( 'Add New Stripe Account', 'give' ); ?></h3>
```
Use `_e()`

bradp commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -676,105 +687,195 @@ public function register_advanced_settings( $settings ) {
+								<tr valign="top" class="give-stripe-account-type-connect">
+									<th scope="row" class="titledesc">
+										<label for="stripe_connect_button">
+											<?php echo __( 'Stripe Connection', 'give' ); ?>
```
Use `_e()`

bradp commented on includes/gateways/stripe/includes/class-give-stripe-webhooks.php:
```diff
@@ -63,16 +63,23 @@ public function listen() {
 			}
 
+			// Retrieve the request's body and parse it as JSON.
+			$body  = @file_get_contents( 'php://input' );
```
Silencing all errors isn't the best idea, let's change this to a try & catch so we can process the errors and actually surface something that is wrong or bail out.

bradp commented on includes/gateways/stripe/includes/class-give-stripe-webhooks.php:
```diff
@@ -63,16 +63,23 @@ public function listen() {
+			 * @Todo Utilize idempotency key on every Stripe request for better stability.
+			 */
+			$metadata    = (array) $event->data->object->metadata;
+			$donation_id = ! empty( $metadata['Donation Post ID'] ) ? $metadata['Donation Post ID'] : 0;
```
Needs localization function.

mehul0810 replied:
@bradp This doesn't require localization as we are passing the information to Stripe metadata and not storing it in site DB.

bradp commented on includes/gateways/stripe/includes/class-give-stripe-webhooks.php:
```diff
@@ -63,16 +63,23 @@ public function listen() {
+			 * @Todo Utilize idempotency key on every Stripe request for better stability.
+			 */
+			$metadata    = (array) $event->data->object->metadata;
+			$donation_id = ! empty( $metadata['Donation Post ID'] ) ? $metadata['Donation Post ID'] : 0;
```
Missing spaces before/after `'`.

mehul0810 replied:
@bradp I think the static array keys don't require spaces before/after `'`. Can you please confirm the same?

bradp commented on includes/gateways/stripe/includes/filters.php:
```diff
@@ -44,3 +44,26 @@ function give_stripe_get_payment_txn_id_fallback( $payment_id ) {
+ * @return mixed
+ */
+function __give_stripe_form_add_credentials( $form_html_tags, $form ) {
+
```
Extra line break.

bradp commented on includes/gateways/stripe/includes/payment-methods/class-give-stripe-becs.php:
```diff
@@ -16,22 +16,22 @@
  * Check for Give_Stripe_Becs existence.
  *
- * @since 2.6.3
+ * @since 2.7.0
```
If this has existed since 2.6.3, then this shouldn't be updated.

mehul0810 replied:
@bradp This BECS Direct Debit related functionality is not released and will get released as a part of `2.7.0`. So, I have updated it to avoid confusion in the future.

bradp commented on includes/gateways/stripe/includes/payment-methods/class-give-stripe-becs.php:
```diff
@@ -127,7 +127,7 @@ public function add_mandate_form( $form_id, $args, $echo = true ) {
 							<?php echo __( 'Bank Account', 'give' ); ?>
 							<span class="give-required-indicator">*</span>
-							<span class="give-tooltip give-icon give-icon-question" data-tooltip="The (typically) 16 digits on the front of your credit card."></span>
+							<span class="give-tooltip give-icon give-icon-question" data-tooltip="<?php echo __( 'BSB Number and Account Number of your bank account.', 'give' );?>"></span>
```
Use `e_()`

ravinderk replied:
@mehul0810 use `esc_attr_e` instead

@bradp Let us know if disagree

bradp commented on includes/gateways/stripe/includes/payment-methods/class-give-stripe-becs.php:
```diff
@@ -127,7 +127,7 @@ public function add_mandate_form( $form_id, $args, $echo = true ) {
 							<?php echo __( 'Bank Account', 'give' ); ?>
 							<span class="give-required-indicator">*</span>
-							<span class="give-tooltip give-icon give-icon-question" data-tooltip="The (typically) 16 digits on the front of your credit card."></span>
+							<span class="give-tooltip give-icon give-icon-question" data-tooltip="<?php echo __( 'BSB Number and Account Number of your bank account.', 'give' );?>"></span>
```
Needs escaping.

bradp commented on includes/gateways/stripe/includes/payment-methods/class-give-stripe-becs.php:
```diff
@@ -192,7 +192,7 @@ class="give-stripe-becs-bank-account-field give-stripe-cc-field"
 		 *
 		 * @return void
-		 * @since  2.6.3
+		 * @since  2.7.0
```
If this has existed since 2.6.3, this shouldn't be updated.

mehul0810 replied:
@bradp This BECS Direct Debit related functionality is not released and will get released as a part of 2.7.0. So, I have updated it to avoid confusion in the future.

bradp commented on includes/gateways/stripe/includes/payment-methods/class-give-stripe-card.php:
```diff
@@ -290,7 +290,7 @@ public function process_payment( $donation_data ) {
 						sprintf(
 							/* translators: %s Donation Data */
-							__( 'Customer creation failed while processing the donation. Details: %s', 'give' ),
+							__( 'Unable to get Stripe Customer ID while processing donation. Details: %s', 'give' ),
```
Needs escaping.

bradp commented on tests/unit-tests/stripe/tests-give-stripe-admin-helpers.php:
```diff
@@ -30,32 +30,6 @@ public function setUp() {
-	 *
-	 * @return void
-	 */
-	public function test_give_stripe_is_connected() {
```
Any reason for removing this test?

mehul0810 replied:
@bradp Because we no longer use this fn and is deprecated

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +430,102 @@ function give_stripe_show_currency_notice() {
+
+		// Bailout, if Account Name already exists.
+		if ( in_array( $new_account_slug, $account_keys, true ) ) {
+			wp_send_json_error( [ 'message' => __( 'This account name is already in use. Please enter a different account name.', 'give' ) ] );
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +430,102 @@ function give_stripe_show_currency_notice() {
+		];
+		wp_send_json_success( $success_args );
+	} else {
+		wp_send_json_error( [ 'message' => __( 'Unable to update account name. Please contact support.', 'give' ) ] );
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+
+	$settings['stripe_form_account_options'] = array(
+		'id'        => 'stripe_form_account_options',
+		'title'     => __( 'Manage Accounts', 'give' ),
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+		),
+		'fields'    => array(
+			array(
+				'name'    => __( 'Account Options', 'give' ),
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+				'type'    => 'radio_inline',
+				'default' => 'disabled',
+				'options' => array(
+					'disabled' => __( 'Global', 'give' ),
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+				'default' => 'disabled',
+				'options' => array(
+					'disabled' => __( 'Global', 'give' ),
+					'enabled'  => __( 'Customize', 'give' ),
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+				),
+			),
+			array(
+				'name'          => __( 'Stripe Accounts', 'give' ),
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+				'title' => sprintf(
+					'<a href="%1$s">%2$s</a> %3$s',
+					admin_url( 'edit.php?post_type=give_forms&page=give-settings&tab=gateways&section=stripe-settings' ),
+					__( 'Click here', 'give' ),
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+					'<a href="%1$s">%2$s</a> %3$s',
+					admin_url( 'edit.php?post_type=give_forms&page=give-settings&tab=gateways&section=stripe-settings' ),
+					__( 'Click here', 'give' ),
+					__( 'to add new Stripe account.', 'give' )
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-helpers.php:
```diff
@@ -204,3 +213,22 @@ function give_stripe_connect_delete_options() {
+function give_stripe_connection_type_names( $slug = 'connect' ) {
+
+	$names = [
+		'manual'  => __( 'API Keys', 'give' ),
```
@mehul0810 use `esc_html`

ravinderk commented on includes/gateways/stripe/includes/admin/admin-helpers.php:
```diff
@@ -204,3 +213,22 @@ function give_stripe_connect_delete_options() {
+
+	$names = [
+		'manual'  => __( 'API Keys', 'give' ),
+		'connect' => __( 'Stripe Connect', 'give' ),
```
@mehul0810 use `esc_html`

mehul0810 commented:
@DevinWalker Please review the UI for account name and email for the Stripe account. The account name and email will be fetched only for Stripe accounts connected via the Connect button. So, the "Edit" link will not appear for connected accounts. For manual API keys, the Account Name will be dynamically generated and the user can change the name.

![image](https://user-images.githubusercontent.com/1852711/81307778-a96dcf80-909e-11ea-8ee7-11111478c6c1.png)

Let me know if you have any suggestions.

mehul0810 commented on includes/gateways/stripe/includes/class-give-stripe-webhooks.php:
```diff
@@ -63,16 +63,23 @@ public function listen() {
 			}
 
+			// Retrieve the request's body and parse it as JSON.
+			$body  = @file_get_contents( 'php://input' );
```
@bradp This is legacy code that we are using from the start and not introduced as a part of the improvement of this issue. So, I would suggest creating a separate issue for this.

DevinWalker commented:
@mehul0810 thanks for the update. Can you add some left and right padding to the email field? In your screenshot it's too close to the elements surrounding it. Give it some room to breathe!

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -282,6 +346,12 @@ function give_stripe_show_connect_banner() {
 
+	// Additional Check: For multiple accounts management.
+	$all_accounts = give_stripe_get_all_accounts();
+	if ( is_array( $all_accounts ) && count( $all_accounts ) > 0 ) {
```
@mehul0810 does not need to use `is_array` and `count`. 
```suggestion
	if ( $all_accounts ) {
```
Update `give_stripe_get_all_accounts` function  to return result only in `array` format.

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +458,102 @@ function give_stripe_show_currency_notice() {
+function give_stripe_set_account_default() {
+
+	$post_data    = give_clean( $_POST );
+	$account_slug = ! empty( $post_data['account_slug'] ) ? give_clean( $post_data['account_slug'] ) : false;
```
@mehul0810 remove extra function call for `give_clean`.
```suggestion
	$account_slug = ! empty( $post_data['account_slug'] ) ? $post_data['account_slug'] : false;
```

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +458,102 @@ function give_stripe_show_currency_notice() {
+
+	if ( $is_updated ) {
+		wp_send_json_success();
+	} else {
```
@mehul0810 No need of `else` condition because `wp_send_json_success` calls `wp_die` internally.

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +458,102 @@ function give_stripe_show_currency_notice() {
+		wp_send_json_error();
+	}
+
+	give_die();
```
@mehul0810 No need of `give_die` condition because `wp_send_json_success` calls `wp_die` internally.

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +458,102 @@ function give_stripe_show_currency_notice() {
+			'slug'    => $new_account_slug,
+		];
+		wp_send_json_success( $success_args );
+	} else {
```
@mehul0810 No need of `give_die` and `clean` condition because wp_send_json_success calls wp_die internally.

ravinderk commented on includes/gateways/stripe/includes/admin/admin-filters.php:
```diff
@@ -34,4 +34,66 @@ function give_stripe_link_transaction_id( $transaction_id, $payment_id ) {
+	$form_account         = give_is_setting_enabled( give_get_meta( $form_id, 'give_stripe_per_form_accounts', true ) );
+	$account_options      = give_stripe_get_account_options();
+	$account_options_keys = array_keys( $account_options );
+	$first_account_key    = is_array( $account_options_keys ) && count( $account_options_keys ) > 0 ? $account_options_keys[0] : [];
```
@mehul0810 we can update this code. `array_keys ` will always return `array`.
```suggestion
	$first_account_key    = $account_options_keys ? $account_options_keys[0] : [];
```

ravinderk commented on includes/gateways/stripe/includes/admin/admin-helpers.php:
```diff
@@ -204,3 +179,22 @@ function give_stripe_connect_delete_options() {
+ *
+ * @return string
+ */
+function give_stripe_connection_type_names( $slug = 'connect' ) {
```
@mehul0810 Are we using this function get label for connecting type if yes then function name can be different then existing.
```suggestion
function give_stripe_connection_type_name( $slug = 'connect' ) {
```

ravinderk commented on includes/gateways/stripe/includes/admin/admin-helpers.php:
```diff
@@ -204,3 +179,22 @@ function give_stripe_connect_delete_options() {
 }
+
+/**
+ * This helper function is used to convert slug into names of Stripe connection.
```
@mehul0810 this function return a single name, so fix typo in the description.
```suggestion
 * This helper function is used to convert slug into the name of Stripe connection.
```

ravinderk commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -76,7 +76,15 @@ public function __construct() {
 			add_filter( 'give_get_sections_advanced', array( $this, 'register_advanced_sections' ) );
 			add_filter( 'give_get_settings_advanced', array( $this, 'register_advanced_settings' ), 10, 1 );
-			add_action( 'give_admin_field_stripe_connect', array( $this, 'stripe_connect_field' ), 10, 2 );
+			add_action(
```
@mehul0810 fix formatting and make it inline as others?

ravinderk commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -686,95 +703,198 @@ public function stripe_modal_checkout_status( $status = 'enabled' ) {
+					<div id="give-stripe-account-manager-errors"></div>
+					<div id="give-stripe-account-manager-description">
+						<h2><?php esc_html_e( 'Manage Your Stripe Accounts', 'give' ); ?></h2>
+						<p class="give-field-description"><?php _e( 'In this section you can connect one or multiple Stripe accounts. All donation forms will use the "Default Account" attached unless you specify otherwise in the specific donation form\'s settings. Connecting multiple accounts allows you to create donation campaigns for specific accounts rather than just one.', 'give' ); ?></p>
```
@mehul0810 can we escape text?

ravinderk commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -686,95 +703,198 @@ public function stripe_modal_checkout_status( $status = 'enabled' ) {
+									<div id="give-stripe-<?php echo $slug; ?>" class="give-stripe-account-manager-list-item">
+										<div class="give-stripe-account-name-wrap">
+											<span class="give-stripe-account-name">
+												<?php _e( $account_name, 'give' ); ?>
```
@mehul0810 translation function does not work with dynamic string. can we is this?

ravinderk commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -686,95 +703,198 @@ public function stripe_modal_checkout_status( $status = 'enabled' ) {
+												<?php _e( $account_name, 'give' ); ?>
+											</span>
+											<span class="give-field-description give-stripe-account-email">
+												<?php _e( $account_email, 'give' ); ?>
```
@mehul0810 translation function does not work with dynamic string. can we is this?

ravinderk commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -686,95 +703,198 @@ public function stripe_modal_checkout_status( $status = 'enabled' ) {
+											if (
+												$slug !== $default_account ||
+												(
+													is_array( $stripe_accounts ) &&
```
@mehul0810 I already added a message related to this that `give_stripe_get_all_accounts` should return `array` and in this way, we can clear these extra checking.

ravinderk commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -686,95 +703,198 @@ public function stripe_modal_checkout_status( $status = 'enabled' ) {
+														href="<?php echo $disconnect_url; ?>"
+														data-disconnect-message="<?php echo $disconnect_message; ?>"
+													>
+																								<?php esc_html_e( 'Disconnect', 'give' ); ?>
```
@mehul0810 can we fix formatting?

ravinderk commented on includes/gateways/stripe/includes/admin/class-give-stripe-admin-settings.php:
```diff
@@ -686,95 +703,198 @@ public function stripe_modal_checkout_status( $status = 'enabled' ) {
+										</label>
+									</th>
+									<td class="give-forminp">
+										<?php echo give_stripe_connect_button(); ?>
```
@mehul0810 I noticed that `$connected` is an unused variable in `give_stripe_connect_button` function. Can we remove this?

ravinderk commented on includes/gateways/stripe/includes/class-give-stripe-gateway.php:
```diff
@@ -554,15 +542,16 @@ public function verify_payment( $payment_id, $stripe_customer_id, $charge ) {
 		 */
-		public function prepare_metadata( $donation_id = 0 ) {
-			return give_stripe_prepare_metadata( $donation_id );
+		public function prepare_metadata( $donation_id, $donation_data = array() ) {
```
@mehul0810 can we use short array tag `[]`

ravinderk commented on includes/gateways/stripe/includes/give-stripe-scripts.php:
```diff
@@ -40,38 +40,37 @@ function give_stripe_frontend_scripts() {
+			'element_empty_styles'         => give_stripe_get_element_empty_styles(),
+			'element_invalid_styles'       => give_stripe_get_element_invalid_styles(),
+			'float_labels'                 => give_is_float_labels_enabled(
+				array(
```
@mrhul use short array tag

ravinderk commented on includes/admin/upgrades/upgrade-functions.php:
```diff
@@ -3633,3 +3634,71 @@ function give_v263_upgrades() {
+		give_update_option( '_give_stripe_get_all_accounts', $stripe_accounts );
+
+		// Set option to check that data is migrated or not.
+		give_update_option( '_give_stripe_data_migrated', true );
```
@mehul0810 please do not store unrelated settings to `give_settings`.
and also remove old setting after migration completed successfully.

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +456,96 @@ function give_stripe_show_currency_notice() {
+function give_stripe_disconnect_connect_stripe_account() {
+	$get_data = give_clean( $_GET );
+
+	if ( is_admin() && isset( $get_data['stripe_disconnected'] ) ) {
```
@mehul0810 no need to check `is_admin` on `admin_init` hook and please check for user capabilities + nonce other any other user can delete stripe account but just visiting URL. `manual` type connection is easy to find.

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -32,27 +32,83 @@ function give_stripe_connect_save_options() {
+	if ( ! $stripe_accounts ) {
 
+		// Set the connect details of Stripe account as default Stripe account.
+		give_update_option( 'give_stripe_connected', $get_vars['connected'] );
```
Please review usage of these key and update code accordingly as we discussed: https://givewp.slack.com/archives/C0FAGC83C/p1590141922067000

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +456,96 @@ function give_stripe_show_currency_notice() {
+ * @return void
+ */
+function give_stripe_set_account_default() {
+	$post_data    = give_clean( $_POST );
```
@mehul0810 always uses user capability + nonce when integrating with the database. Logic like this can provide easy access to any anonymous.

ravinderk commented on includes/gateways/stripe/includes/admin/admin-actions.php:
```diff
@@ -388,3 +456,96 @@ function give_stripe_show_currency_notice() {
+ * @return void
+ */
+function give_stripe_update_account_name() {
+	$post_data        = give_clean( $_POST );
```
@mehul0810 always uses user capability + nonce when integrating with the database. Logic like this can provide easy access to any anonymous.

ravinderk commented on includes/gateways/stripe/includes/admin/admin-helpers.php:
```diff
@@ -166,24 +101,31 @@ function give_stripe_connect_button() {
  */
-function give_stripe_disconnect_url() {
+function give_stripe_disconnect_url( $account_id = '', $account_name = '' ) {
+	$args = [
```
@mehul0810 I did not notice any nonce in this URL. Can we tighten security here?

ravinderk commented:
@mehul0810 Is this function in use: `give_stripe_connect_delete_options`. As we moved functionality to multi-account, I do not see any benefit of this function and related code. Can we clean it?

Let me if you have any questions.

ravinderk commented:
@mehul I found the following issue.

- [ ] Account information row is not responsive:

![image](https://user-images.githubusercontent.com/1784821/82872664-f8e24580-9f50-11ea-8a30-8ecf7e8f53ee.png)


- [ ] We are not removing `test_publishable_key`:

![image](https://user-images.githubusercontent.com/1784821/82873491-35627100-9f52-11ea-81d3-07e7d3fce69a.png)

- [ ] It is not a good idea to set any the latest account as the default. Please do not make new account a default. Let admin decide that:

![image](https://user-images.githubusercontent.com/1784821/82873792-a7d35100-9f52-11ea-8d6b-187354e13d5b.png)

- [ ] My account name is not appearing in the list:

![image](https://user-images.githubusercontent.com/1784821/82874011-fd0f6280-9f52-11ea-8552-9d751e84f80a.png)

![image](https://user-images.githubusercontent.com/1784821/82874062-10bac900-9f53-11ea-9b90-5f73b432e4d7.png)

![image](https://user-images.githubusercontent.com/1784821/82874098-19ab9a80-9f53-11ea-8c1a-5a32c660167c.png)

![image](https://user-images.githubusercontent.com/1784821/82874190-347e0f00-9f53-11ea-908f-4ef019f9c205.png)

- [ ] I am seeing an incorrect model when disconnecting the manual API key account.
![image](https://user-images.githubusercontent.com/1784821/82892954-f4785580-9f6d-11ea-9b98-9e64255825f9.png)


- [ ] Are we showing any fallback name for an account if the name does not exist?
![image](https://user-images.githubusercontent.com/1784821/82892141-8f703000-9f6c-11ea-97f4-a7554960c654.png)

mehul0810 commented:
@ravinderk Can you test and review this PR again?

DevinWalker commented:
@mehul0810 please resolve the merge conflicts and I'll merge it!
